spec/controllers/challenge_signup_controller_spec.rb

```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+require 'spec_helper'
+
+RSpec.describe ChallengeSignupsController, type: :controller do
+  include LoginMacros
+  include RedirectExpectationHelper
+  let(:user) { create(:user) }
+  let(:signup) do
+    challenge = create(:gift_exchange, :closed)
+    collection = create(:collection, challenge: challenge)
+    signup = build(:challenge_signup, collection_id: collection)
+    signup.collection = collection
+    signup.save
+    signup
+  end
+  let(:collection) { signup.collection }
+  let(:challenge) { signup.collection.challenge }
+  let(:collection_owner) { User.find(collection.all_owners.first.user_id) }
+  let(:signup_owner) { Pseud.find(signup.pseud_id).user }
+
+  let(:open_signup) do
+    challenge = create(:gift_exchange, :open)
+    collection = create(:collection, challenge: challenge)
+    signup = build(:challenge_signup, collection_id: collection)
+    signup.collection = collection
+    signup.save
+    signup
+  end
+  let(:open_collection) { open_signup.collection }
+  let(:open_challenge) { open_signup.collection.challenge }
+  let(:open_signup_owner) { Pseud.find(open_signup.pseud_id).user }
+
+  describe "new" do
+    it "ensures signups are open" do
+      fake_login_known_user(user)
+      get :new, collection_id: collection.name, pseud: user.pseuds.first
+      it_redirects_to_with_error(collection_path(collection), \
+                                 "Sign-up is currently closed: please contact a moderator for help.")
+    end
+  end
+
+  describe "show" do
+    xit "checks that there is a challenge" do
+      fake_login_known_user(collection_owner)
+      get :show, id: 999_999, collection_id: collection.name
+      it_redirects_to_with_error(collection_path(collection), \
+                                 "What sign-up did you want to work on?")
+    end
+
+    it "checks ownership" do
+      fake_login_known_user(user)
+      get :show, id: signup, collection_id: collection.name
+      it_redirects_to_with_error(collection_path(collection), \
+                                 "Sorry, you're not allowed to do that.")
+    end
+  end
+
+  describe "index" do
+    it "checks for the right owner" do
```

Comment:
it redirects and errors if the current user is not allowed to see the specified user's sign-ups